### PR TITLE
propagate annotations when inlining blocks (#2727)

### PR DESCRIPTION
**Pull request by grg**
Copy annotations when inlining ControlBlock and ParserBlock objects.

Please let me know what you think of this approach to solve #2727 

There's also the open question of whether we should propagate all or a subset of annotations. Any opinions on how to proceed here?

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
You should add your example as a test in testdata/p4_16_samples.
I would also use a policy function to decide which annotations to skip copying. 
The main problem with block annotations is that the compiler does not understand them so they inhibit some optimizations. But ignoring them is not a solution.

**Comment by mbudiu-vmw** [comment]
Also, please add the outputs generated when running your test.

**Comment by grg** [comment]
I've added a mechanism to exclude propagation of specific annotations. Right now I have a static set containing the excluded annotation names in the `Inline` class, plus static accessors. Please let me know if you think this belongs elsewhere.

I've also added a test to verify that annotation propagation, and exclusion of the `@name` annotation from propagation.

**Comment by mbudiu-vmw** [comment]
You have one failure because apparently the output that you checked in for the example does not match (auto-generated action names do not match).
Is this the output you expect? Notice that the parser annotation is not preserved, since the parser is not inlined as a block.
If you are happy with just control annotations this implementation may be what you want.
We should probably discuss this topic in the language design working group.

File: frontends/p4/inlining.h [hunk 1]
```diff
@@ -225,7 +225,9 @@ class GeneralInliner : public AbstractInliner<InlineList, InlineSummary> {
     }
     // controlled visiting order
     const IR::Node* preorder(IR::MethodCallStatement* statement) override;
-    template<class T> void inline_subst(T *caller, IR::IndexedVector<IR::Declaration> T::*locals);
+    template<class T, class U>
+    void inline_subst(T *caller, IR::IndexedVector<IR::Declaration> T::*locals,
```
Comment (mbudiu-vmw) [inline frontends/p4/inlining.h:229]: Please document the U argument.

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

File: frontends/p4/inlining.h (hunk 1 above)
Comment (grg) [reply frontends/p4/inlining.h]: I've replaced `T` and `U` with `P4Block` and `P4BlockType`, matching the definition in the cpp file. I've also copied the comment from the cpp file into the header.

**Comment by grg** [comment]
> You have one failure because apparently the output that you checked in for the example does not match (auto-generated action names do not match).

I must have missed rerunning the test after my last change to the input file. Corrected.


> Is this the output you expect? Notice that the parser annotation is not preserved, since the parser is not inlined as a block.
> If you are happy with just control annotations this implementation may be what you want.
> We should probably discuss this topic in the language design working group.

The parser annotations (`@my_anno_1` and `@my_anno_2`) propagate through to the midend P4 file. I think this is all I need to get them to the Barefoot compiler 🤞
I'll double-check this in the next day or two.

**Review by mbudiu-vmw — APPROVED** [review]

**Comment by mbudiu-vmw** [comment]
Should we merge this?

**Pull request merged**